### PR TITLE
Converted the `switch` statements to `XSIMD_IF_CONSTEXPR`

### LINKED ISSUE #779: Compile-time dispatch instead of switch-case
As I pointed out in one of my messages in #775, there are many cases like [this](https://github.com/xtensor-stack/xsimd/blob/c2728632bdfecf1dfc004e7ef9ee908003169948/include/xsimd/arch/xsimd_sse2.hpp#L52), where a switch-case is employed to make the decision of which intrinsic to call, even though this decision could be done at compile-time.

As part of our program to improve debug build performance, as @marzer suggested, and for improving performance overall, this should be changed to using some `std::enable_if` type thing.

### PR DESCRIPTION
**Pull request by wermos**
Fixed #779, as per the discussions.

### CODE REVIEW

**Comment by wermos** [comment]
Hmm, it seems like two of the Azure builds timed out.

**Comment by JohanMabille** [comment]
Indeed, the page providing the Intel SDE was temporarily unavailable. I've restarted the job.
Thanks for all the work on this PR!

File: include/xsimd/xsimd.hpp [hunk 1]
```diff
@@ -35,6 +35,21 @@
 #define XSIMD_NO_DISCARD
 #endif
 
+#ifdef __cpp_if_constexpr
+// this means that the compiler supports the `if constexpr` construct
+#define XSIMD_IF if constexpr
+#endif
+
+#if !defined(XSIMD_IF) && __cplusplus >= 201703L
```
Comment (serge-sans-paille) [inline include/xsimd/xsimd.hpp:43]: Should be and `#elif  __cplusplus >= 201703L`

**Comment by serge-sans-paille** [comment]
I'd like to suggest replacing `XSIMD_IF(cond)` by `if XSIMD_CONSTEXPR (cond)` which is more verbose but more explicit too. Thoughts?

**Comment by marzer** [comment]
> I'd like to suggest replacing XSIMD_IF(cond) by if XSIMD_CONSTEXPR (cond) which is more verbose but more explicit too. Thoughts?

That leaves some ambiguity - `constexpr` has been around since C++11, but `if constexpr` specifically since C++17. You would want to call it something more specific so it doesn't look like an alias for just `constexpr`, e.g.

```cpp
#if defined(__cpp_if_constexpr) || __cplusplus >= 201703L
	#define XSIMD_CONSTEXPR_IF constexpr
#else
	#define XSIMD_CONSTEXPR_IF
#endif

// ...

if XSIMD_CONSTEXPR_IF (sizeof(Foo) == 8)
{
	// ...
}
```

Which, while eliminating the ambiguity, is redundant since now it says "if" twice.

File: include/xsimd/xsimd.hpp (hunk 1 above)
Comment (wermos) [reply include/xsimd/xsimd.hpp:43]: I was simply following the way we defined `XSIMD_NO_DISCARD`, via a bunch of cascading checks. But if you still want me to change it, I will do so.

**Comment by JohanMabille** [comment]
Thanks!

**Pull request merged**